File: Application/Classes/Controllers/TabProfiles/Deals/DealsController.m
//
//  DealsController.m
//  CoAssets
//
//  Created by Macintosh HD on 10/2/15.
//  Copyright (c) 2015 Sanyi. All rights reserved.
//

#import "DealsController.h"
#import "COPositive&NagitiveButton.h"
#import "HomeListViewController.h"
#import "WSGetDealProfileRequest.h"
#import "WSURLSessionManager+DealProfile.h"
#import "DealTableViewCell.h"
#import "CODealProfileModel.h"
#import "COLoginManager.h"

@interface DealsController ()<UITableViewDelegate, UITableViewDataSource>
{
    __weak IBOutlet UISegmentedControl *_segment;
    __weak IBOutlet UILabel *_introductionLabel;
    __weak IBOutlet UIView *_contentView;
    __weak IBOutlet UITableView *_tableView;
    __weak IBOutlet UILabel *_greatLabel;
    __weak IBOutlet COPositive_NagitiveButton *__greatButton;
    __weak IBOutlet UIButton *_actionsegment;
}

@property (nonatomic, strong) NSArray *dataArray;
@property (nonatomic, strong) CODealProfileModel *dealModel;

@end

@implementation DealsController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self _setUpUI];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self _callGetDealList];
}

#pragma mark - Getter

- (NSArray *)dataArray {
    if (_dataArray) {
        return _dataArray;
    }
    _dataArray = [[NSArray alloc] init];
    switch (_segment.selectedSegmentIndex) {
        case CODealsStypeOngoing:
            _introductionLabel.text = m_string(@"ON_GOING");
            _dataArray = self.dealModel.dealOngoingModel;
            break;
        case CODealsStypeCompleted:
            _introductionLabel.text = m_string(@"COMPLETED");
            _dataArray = self.dealModel.dealCompleteModel;
            break;
        case CODealsStypeFunded:
            _introductionLabel.text = m_string(@"FUNDED");
            _dataArray = self.dealModel.dealFundedModel;
            break;
    }
    return _dataArray;
}

#pragma mark - Private

- (void)_setUpUI {
    self.navigationItem.title = m_string(@"DEALS");
    [_segment setSelectedSegmentIndex:0];
    [_segment setTitleTextAttributes:@{NSFontAttributeName:[UIFont fontWithName:@"Raleway-Regular" size:15]} forState:UIControlStateNormal];
    [self _setupTableView];
    [self _setTilerLabelAndButton];
    
}

- (void)_setupTableView {
    _tableView.delegate = self;
    _tableView.dataSource = self;
    _tableView.tableFooterView = [UIView new];
    [_tableView registerNib:[UINib nibWithNibName:@"DealTableViewCell" bundle:nil] forCellReuseIdentifier:@"DealTableViewCell"];
}

- (void)_setTilerLabelAndButton {
    _greatLabel.text = m_string(@"TITLE_LABEL");
    [__greatButton setTitle:m_string(@"TITLE_BUTTON") forState:UIControlStateNormal];
}

- (void)_setTilerLabelAndButton:(BOOL)hidden {
    [_tableView setHidden:hidden];
    _actionsegment.hidden = !hidden;
    _greatLabel.hidden = !hidden;
}

- (void)_reloadData {
    self.dataArray = nil;
    [_tableView reloadData];
    BOOL hiden = self.dataArray.count >0;
    [self _setTilerLabelAndButton:!hiden];
}

#pragma mark - Action
- (IBAction)__actionSegment:(id)sender {
    _segment = (UISegmentedControl*)sender;
    [self _reloadData];
}

- (IBAction)__actionButtonGetStart:(id)sender {
    BaseTabBarController *base = [kAppDelegate baseTabBarController];
    [base setSelectedIndex:0];
}

- (void)_callGetDealList {
    [UIHelper showLoadingInView:[kAppDelegate window]];
    NSMutableDictionary *dic = [[NSMutableDictionary alloc]init];
    NSString *device_token = [kUserDefaults objectForKey:KEY_DEVICE_TOKEN];
    if (device_token == nil) {
        device_token = @"debug_Simulator";
    }
    [dic setObject:device_token forKey:device_token_dic];
    [dic setObject:device_type forKey:device_type_dic];
    [dic setObject:application_name forKey:application_name_dic];
    
    [[WSURLSessionManager shared] wsGetDealRequest:dic handler:^(id responseObject, NSURLResponse *response, NSError *error) {
        if (!error && responseObject != nil) {
            self.dealModel = (CODealProfileModel *)responseObject;
            self.dataArray = nil;
            [[NSOperationQueue mainQueue]addOperationWithBlock:^{
                [self _reloadData];
            }];
        } else {
            [ErrorManager showError:error];
        }
        [UIHelper hideLoadingFromView:[kAppDelegate window]];
    }];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.dataArray.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    DealTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"DealTableViewCell" forIndexPath:indexPath];
    cell.deal = [self.dataArray objectAtIndex:indexPath.row];
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 185;
}

@end

File: CoAssetsApps/Targets/Investor/InvestorCore/Classes/Controllers/ProfileFlow/Deals/DealsController.swift
//
//  DealsController.swift
//  CoAssetsApps
//
//  Created by Macintosh HD on 3/4/16.
//  Copyright © 2016 TruongVO07. All rights reserved.
//

import UIKit

class DealsController: BaseViewController {
    
    @IBOutlet weak private var segmentController: UISegmentedControl!
    @IBOutlet weak private var introductionLabel: BaseLabel!
    @IBOutlet weak private var contentView: UIView!
    @IBOutlet weak private var tableView: UITableView!
    @IBOutlet weak private var greatLabel: BaseLabel!
    @IBOutlet weak private var greatButton: COButtonIconDeal!
    @IBOutlet weak private var actionSegment: BaseButton!
    
    var investorDealData: COInvestorDealModel?
    var selectedSegment: Int?
    var isHighLighted: Bool = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        callAPIGetInvestorDeal()
    }
    
    //setupUI
    private func setupUI() {
        title = m_local_string("DEALS")
        segmentController.selectedSegmentIndex = 0
        (segmentController.subviews[1] as UIView).tintColor = UIColor.blackColor()
        (segmentController.subviews[0] as UIView).tintColor = UIColor.blackColor()
        segmentController.setTitle(m_local_string("Ongoing"), forSegmentAtIndex: 0)
        segmentController.setTitle(m_local_string("Funded"), forSegmentAtIndex: 1)
        segmentController.setTitle(m_local_string("Completed"), forSegmentAtIndex: 2)
        setupTableview()
        setTilerLabelAndButton()
    }
    
}

//MARK: Setter, Getter
extension DealsController {
    var dataDealArray: [COInvestorDealDetailModel] {
        set {
            
        }
        get {
            var array = [COInvestorDealDetailModel]()
            if segmentController.selectedSegmentIndex == 0 {
                if let investorDealModel = investorDealData {
                    array = investorDealModel.ongoingArray
                }
                introductionLabel.text = m_local_string("ON_GOING")
            } else if segmentController.selectedSegmentIndex == 1 {
                if let investorDealModel = investorDealData {
                    array = investorDealModel.fundedArray
                }
                introductionLabel.text = m_local_string("FUNDED")
            } else {
                if let investorDealModel = investorDealData {
                    array = investorDealModel.completedArray
                }
                introductionLabel.text = m_local_string("COMPLETED")
            }
            return array
        }
    }
}

//MARK: Private
extension DealsController {
    private func setupTableview() {
        tableView.dataSource = self
        tableView.delegate   = self
        tableView.registerNib(UINib(nibName: DealsCell.className, bundle: nil),
            forCellReuseIdentifier: DealsCell.className)
        tableView.tableFooterView = UIView()
        setTitlerLabelAndButton(dataDealArray.isEmpty)
    }
    
    private func setTilerLabelAndButton() {
        greatLabel.text = m_local_string("TITLE_LABEL")
        greatButton.title = m_local_string("TITLE_BUTTON")
    }
    
    private func setTitlerLabelAndButton(hidden: Bool) {
        tableView.hidden = hidden
        actionSegment.hidden = !hidden
        greatLabel.hidden = !hidden
        
    }
}

//MARK: ReloadData
extension DealsController {
    private func reloadDataDeal() {
        if NSThread.isMainThread() {
            updateInMainThread()
        } else {
            NSOperationQueue.mainQueue().addOperationWithBlock { () -> Void in
                self.updateInMainThread()
            }
        }
    }
    
    private func updateInMainThread() {
        self.dataDealArray.removeAll()
        self.tableView.reloadData()
        self.setTitlerLabelAndButton(self.dataDealArray.isEmpty)
    }
    
}

//MARK: Action
extension DealsController {
    
    @IBAction func indexChanged(sender: UISegmentedControl) {
        segmentController = sender
        reloadDataDeal()
        let sortedViews = sender.subviews.sort({ $0.frame.origin.x < $1.frame.origin.x })
        for (index, view) in sortedViews.enumerate() {
            if index == sender.selectedSegmentIndex {
                view.tintColor = AppDefine.AppColor.CORedColor
            } else {
                view.tintColor = UIColor.blackColor()
            }
        }
    }
    
    @IBAction func actionButtonGetStart(sender: COButtonIconDeal) {
        FlowManager.shared.openHomeFlow()
    }
}

//MARK: UItableviewDatasource
extension DealsController: UITableViewDataSource {
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return dataDealArray.count
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        if let cell = tableView.dequeueReusableCellWithIdentifier(DealsCell.className,
            forIndexPath: indexPath) as? DealsCell {
                cell.dealDetail = dataDealArray[indexPath.row]
                return cell
        } else {
            return UITableViewCell()
        }
    }
}

//MARK: UITableviewDelegate
extension DealsController: UITableViewDelegate {
    func tableView(tableView: UITableView, estimatedHeightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        return UITableViewAutomaticDimension
    }
}

//MARK: API
extension DealsController {
    private func callAPIGetInvestorDeal() {
        let deal = ProfileService()
        UIHelper.showLoadingInView(m_local_string("please_wait"))
        deal.getInvestorDeal { (data, error) -> Void in
            UIHelper.hideLoadingFromView()
            if let _ = error {
                UIHelper.showError(error, actionButton: nil)
            } else {
                if let myData = data as? COInvestorDealModel {
                    self.investorDealData = myData
                    NSOperationQueue.mainQueue().addOperationWithBlock({ () -> Void in
                        self.reloadDataDeal()
                    })
                }
            }
        }
    }
    
}
